src/OmniSharp.Roslyn.CSharp/Workers/Diagnostics/CSharpDiagnosticService.cs

```diff
@@ -1,152 +1,263 @@
+﻿using System;
 using System.Collections.Concurrent;
 using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.Composition;
 using System.Linq;
+using System.Reflection;
+using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Options;
 using Microsoft.Extensions.Logging;
 using OmniSharp.Helpers;
 using OmniSharp.Models.Diagnostics;
-using OmniSharp.Roslyn;
+using OmniSharp.Options;
+using OmniSharp.Services;
 
-namespace OmniSharp.Workers.Diagnostics
+namespace OmniSharp.Roslyn.CSharp.Services.Diagnostics
 {
-    [Export, Shared]
+    [Shared]
+    [Export(typeof(CSharpDiagnosticService))]
     public class CSharpDiagnosticService
     {
-        private readonly ILogger _logger;
-        private readonly OmniSharpWorkspace _workspace;
-        private readonly object _lock = new object();
+        private readonly ILogger<CSharpDiagnosticService> _logger;
```

Comment:
Can you put the new diagnostics service implementation behind the flag? You might have to turn CSharpDiagnosticsService into an aggregator of the "old" and "new" diagnostics services implementations that uses the option to choose between them. Given that we needed to add workarounds to make tests pass in CI, I suspect there are some threading-related issues in this change...

Reply:
I was originally meant to do so that in DI construction phase correct implementation was selected for diagnostic service. However current diagnostic service has very project based interface which doesn't work with file based requests:

`GetCurrentDiagnosticResult(ImmutableArray<ProjectId> projectIds)`

For this reason i implemented 'simpler' but dummier if flag is set use existing extension method to fetch diagnostics. I think more elegant switchable solution is easy to use after api is replaced to use file based analysis api, which i don't know how it works so i don't know what api should look like at this point.